pkg/tenant/topics.go

```diff
@@ -237,24 +254,100 @@ func (s *Server) TopicUpdate(c *gin.Context) {
 		return
 	}
 
+	// Sanity check that the ID in the URL matches the ID in the request body.
+	if topic.ID != topicID.String() {
+		log.Warn().Msg("topic id in request body does not match topic id in URL")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("id in request body does not match id in URL"))
+		return
+	}
+
 	// Verify the topic name exists and return a 400 response if it doesn't.
 	if topic.Name == "" {
 		c.JSON(http.StatusBadRequest, api.ErrorResponse("topic name is required"))
 		return
 	}
 
-	// Get the specified topic from the database and return a 404 response if
-	// it cannot be retrieved.
+	// Topic state is optional but can only be set to readonly.
+	if topic.State != "" && topic.State != pb.TopicTombstone_READONLY.String() {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("topic state can only be set to READONLY"))
+		return
+	}
+
+	// Fetch the topic metadata from the database.
 	var t *db.Topic
-	if t, err = db.RetrieveTopic(c.Request.Context(), topicID); err != nil {
+	if t, err = db.RetrieveTopic(ctx, topicID); err != nil {
+		if errors.Is(err, db.ErrNotFound) {
+			log.Warn().Err(err).Str("topicID", topicID.String()).Msg("topic not found")
+			c.JSON(http.StatusNotFound, api.ErrorResponse("topic not found"))
+			return
+		}
 		log.Error().Err(err).Str("topicID", topicID.String()).Msg("could not retrieve topic")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not update topic"))
+		return
+	}
+
+	// Ensure the new name is valid
+	t.Name = topic.Name
+	if err = t.Validate(); err != nil {
+		log.Warn().Err(err).Msg("could not validate topic update")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(err))
+		return
+	}
+
+	// Verify that the user owns the topic
+	if claims.OrgID != t.OrgID.String() {
+		log.Warn().Str("user_org", claims.OrgID).Str("topic_org", t.OrgID.String()).Msg("user does not own topic")
 		c.JSON(http.StatusNotFound, api.ErrorResponse("topic not found"))
 		return
 	}
 
+	// Only archive mode is available on update
+	if topic.State == pb.TopicTombstone_READONLY.String() {
```

Comment:
If there's no path for unarchiving a topic, should we also protect this endpoint with the confirmation token?

Reply:
There may be a path for unarchiving in the future; I'm not sure though.